src/tree/tree_model.cc

```diff
@@ -618,6 +622,165 @@ std::string RegTree::DumpModel(const FeatureMap& fmap,
+      ForwardN(pos);
+      GetChar(',');
+
+      auto split = common::stoui(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+
+      auto cond = std::stof(raw_str_.substr(cursor_.Pos(), kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(',');
+
+      auto default_left = common::stoui(raw_str_.substr(cursor_.Pos(),
+                                                        kMaxNumLength), &pos);
+      ForwardN(pos);
+      GetChar(']');
+      nodes[i]= {left, right, par, split, cond, static_cast<bool>(default_left)};
+      if (i != n_nodes - 1) {
+        GetChar(',');
+      }
+    }
+    GetChar(']');
+  }
+};
+
+void RegTree::Load(Json const& in) {
+  param.InitAllowUnknown(fromJson(get<Object>(in["tree_param"])));
+  {
+    auto stats_in = get<Raw>(in["stats"]);
+    TreeReader reader(StringView{stats_in.c_str(), stats_in.size()});
+    reader.ParseStat(param.num_nodes, &stats_);
+    CHECK_EQ(param.num_nodes, stats_.size());
+  }
+
+  {
+    auto nodes_in = get<Raw>(in["nodes"]);
+    TreeReader reader(StringView{nodes_in.c_str(), nodes_in.size()});
+    reader.ParseNode(param.num_nodes, &nodes_);
+    CHECK_EQ(param.num_nodes, nodes_.size());
+  }
+}
+
+void RegTree::Save(Json* p_out) const {
+  auto& out = *p_out;
+  CHECK_EQ(param.num_nodes, static_cast<int>(nodes_.size()));
+  CHECK_EQ(param.num_nodes, static_cast<int>(stats_.size()));
+  out["tree_param"] = toJson(param);
+  CHECK_EQ(get<String>(out["tree_param"]["num_nodes"]), std::to_string(param.num_nodes));
+
+  std::string tree_raw;
+  tree_raw.reserve(
+      stats_.size() * std::numeric_limits<Number::Float>::max_digits10 * 2);
+
+  FixedPrecisionStream convertor;
+
+  auto AppendFloat = [&convertor, &tree_raw](Number::Float val, bool end = false) {
+                       convertor << val;
+                       auto const& str = convertor.str();
+                       tree_raw.append(str.c_str(), str.size());
+                       convertor.str("");
+                       if (!end) { tree_raw += ","; }
+                     };
+  auto AppendInt = [&convertor, &tree_raw](int32_t val, bool end = false) {
+                     convertor << val;
+                     auto const& str = convertor.str();
+                     tree_raw.append(str.c_str(), str.size());
+                     convertor.str("");
+                     if (!end) { tree_raw += ","; }
+                   };
+
+  tree_raw += '[';
+  for (size_t i = 0; i < stats_.size(); ++i) {
+    auto const& s = stats_[i];
+    tree_raw += '[';
+    AppendFloat(s.loss_chg, false);
+    AppendFloat(s.sum_hess, false);
+    AppendFloat(s.base_weight, false);
+    AppendInt(s.leaf_child_cnt, true);;
+    tree_raw += ']';
+    if (i != stats_.size() - 1) {
+      tree_raw += ',';
+    }
+  }
+  tree_raw += ']';
+  out["stats"] = JsonRaw(std::move(tree_raw));
+
+  tree_raw.clear();
+  tree_raw.reserve(nodes_.size() * 9 * 2);
```

Comment:
I guess 9 comes from amount of digits one need to save while serializing float. Why do we have 2 here? We have 5 ints and 1 float below

I see that the logic for preallocating buffers for arrays is quite complex and I think this is not the only place we have it. What do you think about extracting this into separate functions?

It is not very critical. It is rather the optimization.